### DISCUSSION TITLE
5G网络下，生成内容会中断

### DISCUSSION THREAD

**Discussion by liuchao0807** [lobehub/lobehub#4692] [General Question]
### 📦 部署环境

Docker

### 📌 软件版本

1.28.5

### 💻 系统环境

iOS

### 🌐 浏览器

Chrome

### 🐛 问题描述

在5G网络条件下，生成长的内容几乎必现中断，导致生成不全
而切到WiFi下，则能顺利生成
除了网络条件，其余条件均一致

### 📷 复现步骤

见视频

https://github.com/user-attachments/assets/e5d5ca2e-76aa-4f63-9179-a153e14f919e

### 🚦 期望结果

任何网络条件下，都能生成完整回复

### 📝 补充信息

会不会和丢包率有关？
或者和超时时间有关？

[a comment by an automated account is not shown]

**Comment by Kac001** [comment]
这肯定是你部署的服务器网络问题

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
> 这肯定是你部署的服务器网络问题

额，我知道肯定和网络有关，但是不知道你看我的实验视频没有，唯一的变量是我**把手机的5G网络，变成了手机连wifi**
服务器的网络完全没变啊

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
> 这肯定是你部署的服务器网络问题

变化的是客户端的网络连接

[a comment by an automated account is not shown]

**Comment by Kac001** [comment]
你要明白，wifi和5g走的路由不是一样的，5g的trace下路由看看是不是那个节点出问题。

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
> 你要明白，wifi和5g走的路由不是一样的，5g的trace下路由看看是不是那个节点出问题。

我是搞嵌入式的，对网络这块不是很熟哈
以我的认知，有几个疑问
1. 5G和wifi即使路由不同，但是都可以连接，说明链路没问题
2. 5G网络条件下，只会生成一半，是否说明网络中途有丢包？或者中途连接断了？
3. 客户端和服务端的连接是UDP还是TCP， 如果中途丢包或断连，是否可以续传？

[a comment by an automated account is not shown]

**Comment by Kac001** [comment]
问题在5g到服务器这段链路上，这种大概率是中途有节点延迟高甚至丢包断连，最好试试换运营商或者用4G，应该是用的http tcp协议

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
lobe

> 问题在5g到服务器这段链路上，这种大概率是中途有节点延迟高甚至丢包断连，最好试试换运营商或者用4G，应该是用的http tcp协议

如果是tcp协议，那理论上应该是有重传机制的？除非是超时了？
对于这种丢包或者延时大的网络条件，是否可以增加超时时间，从而增加传输可靠性？

[a comment by an automated account is not shown]

**Comment by Kac001** [comment]
这就要开发者回答这个问题了

[a comment by an automated account is not shown]

**Comment by Sun-drenched** [comment]
应该还是网络问题——不仅是与部署lobechat的服务器之间的网络连接，还有可能和服务器（或者客户端）与API接口（看你有没有开“客户端请求模式”）的网络连接有关。

如果是使用的是二道贩子那的中转接口，可以尝试换一家，有些二道贩子的接口在国内网络环境下不稳定。

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
> 应该还是网络问题——不仅是与部署lobechat的服务器之间的网络连接，还有可能和服务器（或者客户端）与API接口（看你有没有开“客户端请求模式”）的网络连接有关。
> 
> 
> 
> 如果是使用的是二道贩子那的中转接口，可以尝试换一家，有些二道贩子的接口在国内网络环境下不稳定。

没有使用客户端请求模式，所以实际发起请求的都是我的服务器，对api请求来说，在5G和WiFi模式下是一样的网络条件

[a comment by an automated account is not shown]

**Comment by drsanwujiang** [comment]
分别使用电信和联通 5G 网络测试，Lobe Chat 工作正常。

这种情况通常建议考虑服务器的网络问题，例如服务器所在地是否距离太远（位于美国、欧洲等）。可以使用一台电脑连接手机热点，在电脑上使用 tracert 等工具查看一下路由。

最后，此问题应该与 Lobe Chat 本身无关，建议移至讨论区进行讨论。

[a comment by an automated account is not shown]

**Comment by liuchao0807** [comment]
> 分别使用电信和联通 5G 网络测试，Lobe Chat 工作正常。
> 
> 
> 
> 这种情况通常建议考虑服务器的网络问题，例如服务器所在地是否距离太远（位于美国、欧洲等）。可以使用一台电脑连接手机热点，在电脑上使用 tracert 等工具查看一下路由。
> 
> 
> 
> 最后，此问题应该与 Lobe Chat 本身无关，建议移至讨论区进行讨论。

服务器是香港的
提出这个问题的本意并不是说lobe chat本身有问题，而是看是否可以增加一些技术手段，应对稍差一些的网络条件，提高鲁棒性

**Reply by arvinxx** [reply]
理论上浏览器针对 sse 请求应该会做自动重试的，不太确定还需要的鲁棒性是什么？

**Reply by liuchao0807** [reply]
> 理论上浏览器针对 sse 请求应该会做自动重试的，不太确定还需要的鲁棒性是什么？

针对我发的视频这种断连的情况，从代码上可以解释吗，什么情况下会出？比如丢包率达到多少？或者连续丢包多长时间？ 这种断连从技术上可以规避吗？

**Reply by arvinxx** [reply]
代码上似乎无法解释，有没有试过换个手机网络？

因为我不是搞网络出身，所以不知道丢包率、丢包时长这些更加底层的数据。我们现在的通信协议是 SSE https://developer.mozilla.org/zh-CN/docs/Web/API/Server-sent_events/Using_server-sent_events

我们目前的请求会符合标准的 SSE 格式，不过没有加 retry 参数。这是我目前想到的唯一可尝试的方式

> [retry](https://developer.mozilla.org/zh-CN/docs/Web/API/Server-sent_events/Using_server-sent_events#retry)
> 重新连接的时间。如果与服务器的连接丢失，浏览器将等待指定的时间，然后尝试重新连接。这必须是一个整数，以毫秒为单位指定重新连接的时间。如果指定了一个非整数值，该字段将被忽略。

如果你愿意来试的话，我可以告诉你怎么改来测试验证。

**Reply by liuchao0807** [reply]
> 代码上似乎无法解释，有没有试过换个手机网络？
> 
> 
> 
> 因为我不是搞网络出身，所以不知道丢包率、丢包时长这些更加底层的数据。我们现在的通信协议是 SSE https://developer.mozilla.org/zh-CN/docs/Web/API/Server-sent_events/Using_server-sent_events
> 
> 
> 
> 我们目前的请求会符合标准的 SSE 格式，不过没有加 retry 参数。这是我目前想到的唯一可尝试的方式
> 
> 
> 
> > [retry](https://developer.mozilla.org/zh-CN/docs/Web/API/Server-sent_events/Using_server-sent_events#retry)
> 
> > 重新连接的时间。如果与服务器的连接丢失，浏览器将等待指定的时间，然后尝试重新连接。这必须是一个整数，以毫秒为单位指定重新连接的时间。如果指定了一个非整数值，该字段将被忽略。
> 
> 
> 
> 如果你愿意来试的话，我可以告诉你怎么改来测试验证。

我可以试一下，不过我是搞嵌入式的，基本只用过c语言，玩这个项目也只是使用了docker镜像。如何修改并编译代码可能需要你这边指导下

**Reply by arvinxx** [reply]
https://github.com/lobehub/lobe-chat/blob/d3109318c878af8acdc85ba94c1a2661eca57514/src/libs/agent-runtime/utils/streams/protocol.ts#L95-L97

这里是构造 SSE 输出的代码，我理解改成这样可能就行？

```js
controller.enqueue(`data: ${JSON.stringify(data)}\n`);
controller.enqueue(`retry: 300\n\n`);  // 300ms 重试
```

然后本地 dev 的话，需要先装 node 。

```bsh
// 安装 pnpm
1.  npm i -g pnpm 
// 在项目根目录安装依赖
2. pnpm i
// 启动项目
3. npm run dev
```

然后等待项目编译并启动；

如果dev测试没问题，要打 docker 镜像测试的话，执行 `npm run self-hosting:docker-cn@database` 即可打出本地镜像

https://github.com/lobehub/lobe-chat/blob/d3109318c878af8acdc85ba94c1a2661eca57514/package.json#L55-L57

**Reply by liuchao0807** [reply]
> https://github.com/lobehub/lobe-chat/blob/d3109318c878af8acdc85ba94c1a2661eca57514/src/libs/agent-runtime/utils/streams/protocol.ts#L95-L97
> 
> 这里是构造 SSE 输出的代码，我理解改成这样可能就行？
> 
> ```js
> controller.enqueue(`data: ${JSON.stringify(data)}\n`);
> controller.enqueue(`retry: 300\n\n`);  // 300ms 重试
> ```
> 
> 然后本地 dev 的话，需要先装 node 。
> 
> ```
> // 安装 pnpm
> 1.  npm i -g pnpm 
> // 在项目根目录安装依赖
> 2. pnpm i
> // 启动项目
> 3. npm run dev
> ```
> 
> 然后等待项目编译并启动；
> 
> 如果dev测试没问题，要打 docker 镜像测试的话，执行 `npm run self-hosting:docker-cn@database` 即可打出本地镜像
> 
> https://github.com/lobehub/lobe-chat/blob/d3109318c878af8acdc85ba94c1a2661eca57514/package.json#L55-L57

今天花了点时间，做了以下几个实验
1. 本地dev启动项目，未修改的情况下，也无法复现问题
2. 使用非database版本的docker，也不能复现问题
3. 只有使用database版本的docker，才可以复现问题
感觉这个问题不止一个影响因素，我已经用回非database版了，没太多时间折腾。。。
这个问题暂时就到这把，如果你还有什么想法，可以找我复现抓取log

[a comment by an automated account is not shown]
